Mage/src/main/java/mage/abilities/effects/ContinuousEffects.java

```diff
@@ -523,98 +522,134 @@ public ApprovingObject asThough(UUID objectId, AsThoughEffectType type, Ability
+                            || !effect.applies(idToCheck, affectedAbility, ability, game, controllerId)) {
+                        continue;
                     }
-                    keyChoices.put(choiceKey, choiceValue);
                 }
-                Choice choicePermitting = new ChoiceImpl(true);
-                choicePermitting.setMessage("Choose the permitting object");
-                choicePermitting.setKeyChoices(keyChoices);
-                Player player = game.getPlayer(controllerId);
-                player.choose(Outcome.Detriment, choicePermitting, game);
-                for (ApprovingObject approvingObject : possibleApprovingObjects) {
-                    if (approvingObject.getApprovingAbility().getId().toString().equals(choicePermitting.getChoiceKey())) {
-                        return approvingObject;
-                    }
+
+                if (effect.isConsumable() && !game.inCheckPlayableState()) {
+                    consumableApprovingObjects.add(new ApprovingObject(ability, game));
+                } else {
+                    approvingObjects.add(new ApprovingObject(ability, game));
                 }
+                abilityAsThoughEffectMap.put(ability, effect);
             }
+        }
+        // If only checking for playable, return the first available choice.
+        // The choice doesn't matter as long as something makes it playable.
+        if (game.inCheckPlayableState()) {
+            if (!approvingObjects.isEmpty()) {
+                return approvingObjects.iterator().next();
+            } else if (!consumableApprovingObjects.isEmpty()) {
+                return consumableApprovingObjects.iterator().next();
+            } else {
+                return null;
+            }
+        }
 
+        Player controller = game.getPlayer(controllerId);
+        // Each call to effect.applies will keep adding to the alternative costs for the card
+        // Clear them all out and add back only the one that is being used.
+        controller.clearCastSourceIdManaCosts();
+
+        // Combine consumable and non-consumable together since the player must be able to choose from both
+        approvingObjects.addAll(consumableApprovingObjects);
+        if (approvingObjects.isEmpty()) {
+            return null;
+        } else if (approvingObjects.size() == 1) {
+            ApprovingObject approvingObject = approvingObjects.iterator().next();
+            // Get the effect associated with the approving ability
+            AsThoughEffect asThoughEffect = abilityAsThoughEffectMap.get(approvingObject.getApprovingAbility());
+            // Apply it again before returning in order to add any alternative costs with setCastSourceIdWithAlternateMana
+            // E.g. see Bolas's Citadel
+            asThoughEffect.applies(idToCheck, affectedAbility, approvingObject.getApprovingAbility(), game, controllerId);
```

Comment:
This is a hack. Need to re-write the above and store found approving objects in game.getState().setValue so that effect.applies is only called once when the game is not in the checkPlayableState mode.